### PR TITLE
Cancel Order

### PR DESCRIPTION
**Pull request by michael-alade**
**Resolves**
#1238

**Testing**
- Login as admin
- Choose a payment provider
- In the payment provider settings, choose the methods supported by the payment provider.
**Authorize**: This happens when the payment processor receives the payment.
**De-authorize**: This method runs when you want to cancel a payment before a capture has been done.
**Capture**: This method runs when payment is sent to admin.
**Refund**: This allows the admin make either a full/partial refund to the customer.
- Make an order
- Navigate to the orders page
- Click on the order to start processing

- To cancel before approval, click on the multi-function button toggle then select cancel.
1. Click the cancel button on the multi-function button.
2. A modal would be displayed, telling you it's going to do a full refund of the payment made by the customer. You could select either to restock or not restock the product after the cancel.
3. The order would be moved to the completed tab.

- To cancel the order after payment capture.
1. Click on cancel order.
2. A modal would be displayed, telling you it's going to do a full refund of the payment made by the customer. You could select either to restock or not restock the product after the cancel.
3. The order would be moved to the completed tab.

### CODE REVIEW

**Comment by zenweasel** [comment]
@andela-akolawole This PR should be made to `development` not `master`. I will change

**Comment by rymorgan** [comment]
Here's a new flow for how the cancel should work. We want to build a new component, a multi-function button. This component will be used in other places as well so it should be well documented. 
multi-function button select
multi-function button

Once you select cancel, the button will turn red. You click the button and you will get a confirmation.  
multi-function button cancel
multi-function button popover

**Comment by andela-akolawole** [comment]
@rymorgan So a refund is made after clicking either one of the options.

Also, do you think the order tabs should still be the way it is, after adding a canceled tab?

**Comment by zenweasel** [comment]
@andela-akolawole You have a bunch of changes that are just adding blank lines to files not directly affected. Can you remove those please?

File: imports/plugins/core/orders/client/templates/workflow/shippingSummary.js [hunk 1]
```diff
@@ -111,6 +111,16 @@ Template.coreOrderShippingSummary.helpers({
       }
     });
 
+    const canceled = _.every(shipment.items, (shipmentItem) => {
+      for (const fullItem of order.items) {
+        if (fullItem._id === shipmentItem._id) {
+          if (fullItem.workflow) {
+            return fullItem.workflow.status === "coreOrderItemWorkflow/canceled" ? true : false;
```
Comment (andela-akolawole) [inline imports/plugins/core/orders/client/templates/workflow/shippingSummary.js]: :+1

File: imports/plugins/core/ui/client/components/button/buttonSelect.jsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import React, { PropTypes } from "react";
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/components/button/buttonSelect.jsx]: Rename the file from `.jsx` to `.js`. The jsx extension is not necessary.

**Yes `button.jsx` will be renamed to `button.js` someday, just not today :)**

File: imports/plugins/core/ui/client/components/button/index.js [hunk 3]
```diff
@@ -3,3 +3,4 @@ export { default as IconButton } from "./iconButton";
 export { default as EditButton } from "./editButton";
 export { default as VisibilityButton } from "./visibilityButton";
 export { default as Handle } from "./handle";
+export { default as ButtonSelect } from "./buttonSelect.jsx";
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/components/button/index.js]: omit file extension

File: imports/plugins/core/ui/client/containers/buttonSelectContainer.js [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    const nonActiveButtons = buttons.filter(button => {
+      if (button.name !== activeButton) {
+        return button;
+      }
+    });
+    return this.setState({ nonActiveButtons });
+  }
+
+  handleToggle() {
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/containers/buttonSelectContainer.js]: define functions that you would normally need to bind as like so:

```
handleToggle = () => {

}
```

File: imports/plugins/core/ui/client/containers/buttonSelectContainer.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        nonActiveButtons={nonActiveButtons}
+        handleToggle={this.handleToggle}
+        handleButtonChange={this.handleButtonChange}
+        {...this.props}
+      />
+    );
+  }
+}
+
+ButtonSelectContainer.propTypes = {
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/containers/buttonSelectContainer.js]: make this a class property:

```
class ButtonSelectContainer {
  static propTypes = {}
}
```

File: imports/plugins/core/ui/client/containers/buttonSelectContainer.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
+import React, { Component, PropTypes } from "react";
+import classnames from "classnames";
+import { composeWithTracker, merge } from "/lib/api/compose";
+import Button from "../components/button/button.jsx";
+import { ButtonSelect } from "../components/button";
+
+class ButtonSelectContainer extends Component {
+  constructor(props) {
+    super(props);
+    const currentButton = <Button
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/containers/buttonSelectContainer.js]: all logic around `currentButton` should be in the render function, not set on the state. I like to think of state like something serializable, so no JSX, no functions, only strings, objects, arrays, booleans, numbers ( basically like JSON).

File: imports/plugins/core/ui/client/containers/buttonSelectContainer.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
+import React, { Component, PropTypes } from "react";
+import classnames from "classnames";
+import { composeWithTracker, merge } from "/lib/api/compose";
+import Button from "../components/button/button.jsx";
+import { ButtonSelect } from "../components/button";
+
+class ButtonSelectContainer extends Component {
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/containers/buttonSelectContainer.js]: Container components should only contain business logic and no manipulation of the view. The buttonSelect component should take what ever props you pass in, and do all the necessary rendering.

File: imports/plugins/core/ui/client/containers/buttonSelectContainer.js (hunk 6 above)
Comment (andela-akolawole) [inline imports/plugins/core/ui/client/containers/buttonSelectContainer.js]: 👍

File: imports/plugins/core/ui/client/containers/buttonSelectContainer.js (hunk 7 above)
Comment (andela-akolawole) [inline imports/plugins/core/ui/client/containers/buttonSelectContainer.js]: Hey @mikemurray, thanks for the review but I still don't understand what you mean here.

Comment (mikemurray) [inline imports/plugins/core/ui/client/containers/buttonSelectContainer.js]: We have two sets of components in Reaction.

`Container` components and `Presentational` components.

A Presentational component is like the `Select` or `Button` component, which **do not** connect to an external datasource though `composeWithTracker`, `redux`, or some other method. It's a simple component that can be used anywhere and for any purpose and any manipulation of that view is usually done from a parent or `Container` component through props (avoiding `state` unless absolutely necessary).

`Container` components contain all the logic for a single component. The `Container` component usually wraps a `Presentational` component and is to be used when you want a dumb `Presentational` component to do more with data from external sources. For example the `ProductDetailContainer` fetches product, variants, and media and passes them to the `ProductDetail` `Presentational` component. All the `ProductDetail` component does is render the view.

This is good practice in React so you force yourself to make reusable dumb components, and move the control of them higher up in the render tree to a container component / redux store.

**tl;dr**

Container components = business logic, data fetching, data manipulation
View components = accept props, and render a view based on those props

Comment (mikemurray) [inline imports/plugins/core/ui/client/containers/buttonSelectContainer.js]: So basically, most of this component looks presentational, and should be moved into a `Presentational` component.

Everything related to composeWithTracker, etc, should stay.

File: imports/plugins/core/ui/client/components/button/buttonSelect.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    defaultButton = defaultButton[0];
+
+    const defaultBgClassNames = classnames({ "button-select": true, [defaultButton.bgColor]: true });
+
+    const defaultNonActiveButtons = props.buttons.filter(button => {
+      if (button.active === false || button.active === undefined) {
+        return button;
+      }
+    });
+    const currentButton = <Button
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/components/button/buttonSelect.js]: Indentation is a but weird, try something like following. (kinda like how you did the return in the render)
```
const currentButton = (
  <Button
    yourPropsAndStuff
  >
  </Button>
);
```

**Comment by mikemurray** [comment]
Changes look good so far from a code review standpoint, with just some minor indentation issues.

**Comment by zenweasel** [comment]
If I cancel an order before I capture it am I supposed to see "Captured Total"?
cart_completed

**Comment by zenweasel** [comment]
Should we disable entering shipping info and shipping it after the order is cancelled?

**Comment by zenweasel** [comment]
If I cancel an order before capture, the capture button still shows up.

cart_completed_and_ar_json_-_reaction_-____projects_js_reaction_

File: imports/plugins/core/orders/client/templates/workflow/shippingInvoice.js [hunk 9]
```diff
@@ -358,6 +419,29 @@ Template.coreOrderShippingInvoice.helpers({
     return true;
   },
 
+  checkIfCanceled() {
+    const instance = Template.instance();
+    const order = instance.state.get("order");
+    const orderStatus = orderCreditMethod(order).paymentMethod.status;
+
+    return orderStatus !== "canceled" ? false : true;
```
Comment (zenweasel) [inline imports/plugins/core/orders/client/templates/workflow/shippingInvoice.js]: Can't this be simplified to `return orderStatus === "canceled"?

File: server/methods/core/orders.app-test.js [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  function spyOnMethod(method, id) {
+    return sandbox.stub(Meteor.server.method_handlers, `orders/${method}`, function () {
+      check(arguments, [Match.Any]); // to prevent audit_arguments from complaining
+      this.userId = id;
+      return methods[method].apply(this, arguments);
+    });
+  }
+
+  describe("orders/cancelOrder", () => {
+    it("should return an error if user is not admin", (done) => {
```
Comment (zenweasel) [inline server/methods/core/orders.app-test.js]: You don't want to use arrow functions within Mocha tests. https://mochajs.org/#arrow-functions

File: server/methods/core/orders.app-test.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+      spyOnMethod("cancelOrder", order.userId);
+
+      function copyStartCancelOrder() {
+        return Meteor.call("orders/cancelOrder", order, returnToStock);
+      }
+      expect(copyStartCancelOrder).to.throw(Meteor.Error, /Access Denied/);
+      return done();
+    });
+
+    it("should return the product to stock ", (done) => {
```
Comment (zenweasel) [inline server/methods/core/orders.app-test.js]: You only need to use the `done`/`return done()` syntax when doing something async. https://mochajs.org/#asynchronous-code

File: server/methods/core/orders.js [hunk 12]
```diff
@@ -213,6 +213,63 @@ export const methods = {
...
+   */
+  "orders/cancelOrder": function (order, returnToStock) {
+    check(order, Object);
+    check(returnToStock, Boolean);
+
+    if (!Reaction.hasPermission("orders")) {
+      throw new Meteor.Error(403, "Access Denied");
+    }
+
+    if (!returnToStock) ordersInventoryAdjust(order._id);
```
Comment (zenweasel) [inline server/methods/core/orders.js]: While this is syntactically correct, I prefer would us to be consistent and use the full `if/then` syntax

File: server/methods/core/orders.js [hunk 13]
```diff
@@ -213,6 +213,63 @@ export const methods = {
...
+    paymentMethod = Object.assign(paymentMethod, { amount: Number(paymentMethod.amount) });
+    const invoiceTotal = order.billing[0].invoice.total;
+    const shipment = order.shipping[0];
+    const itemIds = shipment.items.map((item) => {
+      return item._id;
+    });
+
+    // refund payment to customer
+    Meteor.call("orders/refunds/create", order._id, paymentMethod, Number(invoiceTotal), err => {
+      if (err) Logger.warn(err);
```
Comment (zenweasel) [inline server/methods/core/orders.js]: Sometimes we're calling `Logger.warn` sometimes `Logger.error`. We should be consistent. Generally we should use `Logger.error` when handling an error, especially from a server-side Meteor method.

File: server/methods/core/orders.js [hunk 14]
```diff
@@ -213,6 +213,63 @@ export const methods = {
...
+    });
+
+    // send notification to user
+    const prefix = Reaction.getShopPrefix();
+    const url = `${prefix}/notifications`;
+    const sms = true;
+    Meteor.call("notification/send", order.userId, "orderCancelled", url, sms, err => {
```
Comment (zenweasel) [inline server/methods/core/orders.js]: Wrapping the `(err)` in parenthesis makes this easier to read and I would like us to stay consistent. Although if there is a good reason for using this syntax I am open.

**Comment by zenweasel** [comment]
A few super minor things and I think this is pretty much ready to go.

File: imports/plugins/core/orders/client/templates/workflow/shippingInvoice.html [hunk 15]
```diff
@@ -37,14 +37,26 @@
           <i class="fa fa-spinner fa-spin {{#unless isRefunding}}hidden{{/unless}}"></i>
         </button>
       </div>
+      {{#if showAfterPaymentCaptured}}
+          <button class="btn btn-danger" type="button" data-event-action="cancelOrder" style="margin-bottom: 10px"
+          >Cancel Order</button>
```
Comment (aaronjudd) [inline imports/plugins/core/orders/client/templates/workflow/shippingInvoice.html]: Needs i18n.

File: imports/plugins/core/orders/client/templates/workflow/shippingInvoice.html [hunk 16]
```diff
@@ -37,14 +37,26 @@
...
+      {{/if}}
+
+      {{/if}}
+      {{#if paymentCanceled}}
+        <span class="badge badge-danger" style="margin-bottom: 10px">Canceled</span>
```
Comment (aaronjudd) [inline imports/plugins/core/orders/client/templates/workflow/shippingInvoice.html]: Needs i18n

File: imports/plugins/core/orders/client/templates/workflow/shippingInvoice.js [hunk 17]
```diff
@@ -82,6 +83,28 @@ Template.coreOrderShippingInvoice.helpers({
   InvoiceContainer() {
     return InvoiceContainer;
   },
+  buttonSelectComponent() {
+    return {
+      component: ButtonSelect,
+      buttons: [
+        {
+          name: i18next.t("order.approveInvoice"),
```
Comment (aaronjudd) [inline imports/plugins/core/orders/client/templates/workflow/shippingInvoice.js]: @mikemurray is this how we apply i18n to a button, or should there be an i18nkey/label as props?

File: imports/plugins/core/orders/client/templates/workflow/shippingSummary.js [hunk 18]
```diff
@@ -111,6 +111,16 @@ Template.coreOrderShippingSummary.helpers({
       }
     });
 
+    const canceled = _.every(shipment.items, (shipmentItem) => {
```
Comment (aaronjudd) [inline imports/plugins/core/orders/client/templates/workflow/shippingSummary.js]: I'd avoid import lodash and using _ if there is an ES6 equivalent.  (array.every prototype?).

File: server/imports/fixtures/orders.js [hunk 19]
```diff
@@ -16,6 +16,10 @@ export function randomProcessor() {
   return _.sample(["Stripe", "Paypal", "Braintree"]);
 }
 
+const itemIdOne = Random.id();
+const itemIdTwo = Random.id();
+
+
```
Comment (aaronjudd) [inline server/imports/fixtures/orders.js]: excess line spacing, lint this..

**Comment by michael-alade** [comment]
@aaronjudd changes have been made to the files you reviewed. I still haven't gotten any reply regarding your question to @mikemurray 

>@mikemurray is this how we apply i18n to a button, or should there be an i18nkey/label as props?

**Comment by mikemurray** [comment]
Buttons have an i18n key for the label. Its 'i18nKeyLabel'

**Comment by mikemurray** [comment]
But unfortunately im not sure about this button component as it seems to wrap the original button component.

**Comment by michael-alade** [comment]
@mikemurray Do you think it's ok to wrap the button's `i18nkeyLabel` with the multi-function button name?

**Comment by mikemurray** [comment]
@michael-alade you should always pass through the props if possible, such as `i18nKeyLabel`

**Comment by aaronjudd** [comment]
Errors seen in console when canceling order, no restock option: 
```
17:54:57.840Z ERROR Reaction: [ACK Failure: Currency is not supported]
  Error: [ACK Failure: Currency is not supported]
      at [object Object].methods.paypalexpress/refund/create (imports/plugins/included/payments-paypal/server/methods/express.js:264:13)
      at packages/check.js:130:16
      at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
```

File: imports/plugins/core/ui/client/components/button/buttonSelect.js [hunk 20]
```diff
@@ -128,6 +128,7 @@ class ButtonSelect extends Component {
           {nonActiveButtons.map((button, key) => {
             return (<div
               className="button-item" key={key}
+              data-i18n={button.i18nKeyLabel}
```
Comment (mikemurray) [inline imports/plugins/core/ui/client/components/button/buttonSelect.js]: - dont use `data-i18n` attributes, they are of a jQuery/blaze legacy which is going away
- fix the indentation
- use a button component

Comment (michael-alade) [inline imports/plugins/core/ui/client/components/button/buttonSelect.js]: @mikemurray I am currently using a button component for the actual multi-function button. 
Do you mean I should use a button component for the dropdown?

**Comment by michael-alade** [comment]
@aaronjudd Could you give the cancel order issue a try with the paypal express payment provider?

**Comment by zenweasel** [comment]
When I try to cancel an order submitted via Braintree I am getting an error:

```
Object {name: "reaction-client", hostname: "localhost:3000", pid: undefined, level: 40, err: Object…}
"Braintree does not allow refunds until transactions are settled. This can take up to 24 hours. Please try again later. [Attempt to refund transaction failed]"
name
```

**Comment by michael-alade** [comment]
@zenweasel I've added error handling for the failing refunds.

File: imports/plugins/included/payments-braintree/lib/collections/schemas/braintree.js [hunk 21]
```diff
@@ -31,6 +31,14 @@ export const BraintreePackageConfig = new SimpleSchema([
...
       label: "Private Key",
       optional: false
+    },
+    "settings.reaction-braintree.support": {
+      type: Array,
+      label: "Payment provider supported methods"
+    },
+    "settings.reaction-braintree.support.$": {
+      type: String,
+      allowedValues: ["authorize", "de-authorize", "capture", "refund"]
```
Comment (zenweasel) [inline imports/plugins/included/payments-braintree/lib/collections/schemas/braintree.js]: This/these need to be standardized either Capitalized or not capitalized. And when displayed to the customer they should be capitalized.

File: imports/plugins/included/payments-example/client/settings/components/exampleSettingsForm.js [hunk 22]
```diff
@@ -1,47 +1,112 @@
...
+              label="authorize"
+              onChange={this.handleCheckBox}
+              name="authorize"
+              checked={this.state.checkbox.authorize}
+            />
+          </div>
+
+          <div>
+            <Checkbox
+              label="de-authorize"
```
Comment (zenweasel) [inline imports/plugins/included/payments-example/client/settings/components/exampleSettingsForm.js]: These need to be capitalized

File: imports/plugins/included/payments-authnet/lib/collections/schemas/package.js [hunk 23]
```diff
@@ -15,6 +15,14 @@ export const AuthNetPackageConfig = new SimpleSchema([
       type: Boolean,
       defaultValue: false
     },
+    "settings.reaction-auth-net.support": {
+      type: Array,
+      label: "Payment provider supported methods"
+    },
+    "settings.reaction-auth-net.support.$": {
+      type: String,
+      allowedValues: ["Authorize", "De-authorize", "Capture", "Refund"]
```
Comment (zenweasel) [inline imports/plugins/included/payments-authnet/lib/collections/schemas/package.js]: These are not showing up enabled

**Comment by zenweasel** [comment]
When I place an order with Authnet and then try to process the order I get this error
```
modules.js?hash=266cbc4…:9822 Warning: Failed prop type: Invalid prop `refunds` of type `boolean` supplied to `InvoiceContainer`, expected `array`.
    in InvoiceContainer (created by Container(InvoiceContainer))
    in Container(InvoiceContainer)
modules.js?hash=266cbc4…:9822 Warning: Failed prop type: Invalid prop `refunds` of type `boolean` supplied to `Invoice`, expected `array`.
    in Invoice (created by InvoiceContainer)
    in InvoiceContainer (created by Container(InvoiceContainer))
    in Container(InvoiceContainer)
```

**Comment by zenweasel** [comment]
If I process an order via Braintree and then cancel it. Then try to click on the order later (and clicking on other orders) I get this order:
```
11:08:35.691Z ERROR Reaction: Cannot read property 'settings' of undefined
  TypeError: Cannot read property 'settings' of undefined
      at getAccountOptions (imports/plugins/included/payments-braintree/server/methods/braintreeApi.js:46:5)
      at getGateway (imports/plugins/included/payments-braintree/server/methods/braintreeApi.js:68:1)
      at Object.BraintreeApi.apiCall.listRefunds (imports/plugins/included/payments-braintree/server/methods/braintreeApi.js:204:3)
      at [object Object].listRefunds (imports/plugins/included/payments-braintree/server/methods/braintreeMethods.js:154:2)
      at packages/check.js:130:16
      at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
      at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check.js:129:41)
      at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1734:18)
      at packages/ddp-server/livedata_server.js:1666:15
      at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
      at packages/ddp-server/livedata_server.js:1664:30
      at [object Object]._.extend.applyAsync (packages/ddp-server/livedata_server.js:1663:12)
      at [object Object]._.extend.apply (packages/ddp-server/livedata_server.js:1611:26)
      at [object Object]._.extend.call (packages/ddp-server/livedata_server.js:1593:17)
      at packages/kadira:flow-router-ssr/server/plugins/ssr_data.js:67:23
      at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
      at Object.Meteor.(anonymous function) [as call] (packages/kadira:flow-router-ssr/server/plugins/ssr_data.js:66:44)
      at [object Object].methods.orders/refunds/list (server/methods/core/orders.js:1072:4)
      at packages/check.js:130:16
      at [object Object]._.extend.withValue (packages/meteor.js:1122:17)
      at Object.exports.Match._failIfArgumentsAreNotAllChecked (packages/check.js:129:41)
      at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1734:18)
```

**Comment by zenweasel** [comment]
This is from my review 18 days ago

> Should we disable entering shipping info and shipping it after the order is cancelled? And doing a refund (it doesn't work, but why not make it clearer?)

It looks like this is still not done?

File: imports/plugins/included/payments-braintree/lib/collections/schemas/braintree.js (hunk 21 above)
Comment (michael-alade) [inline imports/plugins/included/payments-braintree/lib/collections/schemas/braintree.js]: Working on this now.

File: imports/plugins/included/payments-example/client/settings/components/exampleSettingsForm.js (hunk 22 above)
Comment (michael-alade) [inline imports/plugins/included/payments-example/client/settings/components/exampleSettingsForm.js]: I'll do that.
